File: c/lcd.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#define MAX_INPUT_LENGTH 32

const int TABLE[] = { 119, 36, 93, 109, 46, 107, 123, 37, 127, 111 };
const char H_SYM = '-';
const char V_SYM = '|';
const char S_SYM = ' ';

typedef struct {
    int numbers[MAX_INPUT_LENGTH];
    int length;
    int segments;
} LCD;

void lcd_init(LCD* lcd, char *input, int segments)
{
  lcd->length = strlen(input);
  for (int i = 0; i < lcd->length; ++i)
  {
    if (input[i] >= '0' && input[i] <= '9')
    {
        lcd->numbers[i] = input[i] - '0';
    }
  }
  lcd->segments = segments;
}

bool lcd_test(int number, int position)
{
  if (number < 10)
  {
    return (TABLE[number] & (int)(pow(2.0, position))) > 0;
  } else {
    return false;
  }
}

void lcd_draw(LCD* lcd)
{
  int offset = 0;
  for (int row = 0; row < 5; ++row)
  {
    if (row % 2 == 0)
    {
      for (int i = 0; i < lcd->length; ++i)
      {
        char symbol = lcd_test(lcd->numbers[i], row + offset) ? H_SYM : S_SYM;
        char symbols[lcd->segments];
        for (int j = 0; j < lcd->segments; ++j)
        {
          symbols[j] = symbol;
        }
        symbols[lcd->segments] = '\0';
        printf("%c%s%c%c", S_SYM, symbols, S_SYM, S_SYM);
      }
      printf("\n");
    } else {
      for (int i = 0; i < lcd->segments; ++i)
      {
        for (int j = 0; j < lcd->length; ++j)
        {
          for (int column = 0; column < 2; ++column)
          {
            char symbol = lcd_test(lcd->numbers[j], row + column + offset) ? V_SYM : S_SYM;
            int num_spaces = (column == 0 ? lcd->segments : 1);
            char spaces[num_spaces];
            for (int k = 0; k < num_spaces; ++k)
            {
              spaces[k] = S_SYM;
            }
            spaces[num_spaces] = '\0';
            printf("%c%s", symbol, spaces);
          }
        }
        printf("\n");
      }
      ++offset;
    }
  }
}

int main(int argc, char const *argv[])
{
    if (argc > 1) {
        char input[MAX_INPUT_LENGTH];
        int segments = 2;
        if (strcmp(argv[1], "-s") == 0) {
            if (argc == 4) {
                segments = atoi(argv[2]);
                strcpy(input, argv[3]);
            }
        } else {
            strcpy(input, argv[1]);
        }
        LCD *lcd = malloc(sizeof(LCD));
        lcd_init(lcd, input, segments);
        lcd_draw(lcd);
        free(lcd);
    } else {
        printf("Usage: %s [-s SEGMENTS] NUMBER\n", argv[0]);
    }
  return 0;
}

File: rust/src/lcd.rs
use std::str;

static TABLE: [uint, ..10] = [
  0b1110111,
  0b0100100,
  0b1011101,
  0b1101101,
  0b0101110,
  0b1101011,
  0b1111011,
  0b0100101,
  0b1111111,
  0b1101111
];
static H_SYM: &'static str = "-";
static V_SYM: &'static str = "|";
static S_SYM: &'static str = " ";

struct LCD {
    numbers: ~[uint],
    segments: uint
}

impl LCD {

    fn new(input: ~str, segments: uint) -> LCD {
        let mut numbers = ~[];
        for input.iter().advance |c| {
            let option = std::int::from_str(str::from_char(c));
            let n : uint = match option {
                None => 0,
                Some(y) => { y }
            } as uint;
            numbers.push(n);
        }
        LCD { numbers: numbers, segments: segments }
    }

    fn test(&self, number: uint, position: int) -> bool {
        return (TABLE[number] & (0b0000001 << position)) > 0;
    }

    fn draw(&self) {
        let mut offset = 0;
        for std::int::range(0, 5) |row| {
            if (row % 2 == 0) {
                for self.numbers.iter().advance |&number| {
                    let symbol = if (self.test(number, row + offset)) { H_SYM } else { S_SYM };
                    let mut symbols = ~"";
                    for self.segments.times {
                        symbols = symbols + symbol;
                    }
                    print(S_SYM + symbols + S_SYM + S_SYM);
                }
                print("\n");
            } else {
                for self.segments.times {
                    for self.numbers.iter().advance |&number| {
                        for std::int::range(0, 2) |column| {
                            let symbol = if (self.test(number, row + column + offset)) { V_SYM } else { S_SYM };
                            let mut spaces = ~"";
                            let spaces_count = if (column == 0) { self.segments } else { 1 };
                            for spaces_count.times {
                                spaces = spaces + S_SYM;
                            }
                            print(symbol + spaces);
                        }
                    }
                    print("\n");
                }
                offset = offset + 1;
            }
        }
    }
}

fn main() {
    let args = std::os::args();
    if (args.len() > 1) {
        let mut input: ~str;
        let mut segments: uint = 2;
        if (args[1] == ~"-s") {
            match std::int::from_str(args[2]) {
                Some(n) => { segments = n as uint },
                None => {}
            };
            input = args[3];
        } else {
            input = args[1];
        }
        let lcd = LCD::new(input, segments);
        lcd.draw();
    } else {
        println("Usage: LCD [-s SEGMENTS] NUMBER");
    }
}
